### PR TITLE
GH-89812: Add `pathlib._PathBase`

### PR DESCRIPTION
**Pull request by barneygale**
Add private `pathlib._PathBase` class. This will be used by an experimental PyPI package to incubate a `tarfile.TarPath` class.

## Implementation notes

In `pathlib.py`:

- Split `Path` into `_PathBase` and `Path`
- Implement `_PathBase.is_junction()` and `is_mount()` using `stat()`. In the latter case, restore the implementation deleted in #31458
- Implement `_PathBase.resolve()` using `absolute()` and `readlink()`
- Implement `_PathBase._scandir()` using `iterdir()`

In `test_pathlib.py`:

- Add `PathBaseTest` to exercise the `_PathBase` class directly
- Split `PathTest` into `DummyPathTest` and `PathTest`. The former exercises a representative subclass of `_PathBase`.
- Add `DummyPathWithSymlinksTest` to exercise a similar subclass, this time with symlink support.

## Questionable bits

In `pathlib.py`, the `_PathBase.is_junction()` and `_PathBase.resolve()` methods are new code. The latter is very performance-sensitive and may need more work.

## Future work

- Make `pathlib._PathBase` public. There's still at least a couple of things to do first:
  - ~~Make `PurePath._flavour` public~~
  - Figure out what to do with `_PathBase.__hash__()`, `__eq__()`, and other comparison methods.

---

* Issue: gh-89812

### CODE REVIEW

**Comment by barneygale** [comment]
@brettcannon a long time ago in [a PR](https://github.com/python/cpython/pull/31085#issuecomment-1059590608) far, far away, you said:

> My current thinking (while I work through my PR review backlog to reach this PR), is this should be used for `zipfile.Path`. I also wouldn't be opposed to that happening this PR if @barneygale wanted to give that a shot while he waits for a review from me.

I couldn't make this work for `zipfile.Path` due to a couple of fundamental differences in the interface, e.g. the [`root`](https://github.com/jaraco/zipp/issues/95) attribute stores a string in pathlib, but a `ZipFile` instance in `zipfile.Path`. It's possible to make it work with a couple of deprecations/removals, or an entirely new `ZipPath` class.

So in this PR I've chosen `tarfile.TarPath` as the initial target. I think it does a good job of showing the viability of the `pathlib._VirtualPath` interface.

However, I have a bunch of open questions in the PR description about how `TarPath` itself should work. It's not as straightforward as `.zip` files, because tarballs support symlinks and hardlinks, and because `tarfile.TarFile` doesn't support reading and writing the same file.

These questions make me think that `tarfile.TarPath` should first be gestated in a PyPI package before we add it to the standard library. But that leaves a bit of a chicken-and-egg problem with getting `pathlib._VirtualPath` in.

Do you have any advice on the best way forward?

Thanks!

**Comment by brettcannon** [comment]
> These questions make me think that `tarfile.TarPath` should first be gestated in a PyPI package before we add it to the standard library. But that leaves a bit of a chicken-and-egg problem with getting `pathlib._VirtualPath` in.

You have a couple of options:

1. Start a discussion on discuss.python.org for `tarfile` and see if there's clear consensus on what to do.
2. Make `_VirtualPath` as private and do your experimental PyPI project knowing it might be perpetually tied to a specific Python version (and make that clear to users of the package).
3. Go with your gut and hope for the best (and err on the side of leaving things out, especially if they can be added in later). 😅

**Comment by merwok** [comment]
I am unsure about the name VirtualPath!

The way I’ve seen it used, a TarPath class, or SshPath, or S3Path would be examples of virtual filesystems: I give them paths that look like regular filesystem paths and they do custom operations to return info or data.

Their base class is not itself a virtual path, it’s an (abstract?) base class!

**Comment by barneygale** [comment]
Fair point! How about `PathBase`? It squares with the names of base classes in `io`, I think.

**Comment by barneygale** [comment]
Alternatively, `pathlib.abc.VirtualPath`?

**Comment by barneygale** [comment]
I've removed all the `tarfile` changes and renamed the class to `_PathBase` per @merwok. The symlink behaviour isn't clear-cut, so I'll incubate `tarfile.TarPath` in a PyPI package first.

**Comment by merwok** [comment]
Not sure where to discuss this but here goes:
Barney, are you interested in porting changes to one of the pathlib backports?

There is pathlib2 but it seems to have a contribution / leadership issue: https://github.com/jazzband/pathlib2/issues/70
pathlib3x seems active but is a one-person project and adds extensions so not a pure in-place backport.

**Comment by ap--** [comment]
I'd be happy to vendor a backport in universal_pathlib. In fact it's kind of on the roadmap to do this so I don't have to support all individual python versions below 3.12 (or in fact 3.13) separately. 😊

**Comment by merwok** [comment]
Would be a shame to have the code copy-pasted in all the projects in the pathlib ecosystem!

**Comment by barneygale** [comment]
Good idea! I'm happy to publish + maintain a backport package that others (e.g. universal_pathlib folks) can depend on when `python_version < "3.13"`. I'll start work on that once we have something worth backporting, namely a public `pathlib.PathBase`. That may only be weeks away. Does that work for you, @merwok and @ap--?

**Comment by merwok** [comment]
That sounds great.  Presumably the jazzband package would be a good one to update?

File: Lib/pathlib.py
Comment (barneygale) [inline Lib/pathlib.py]: This implementation of `_PathBase.is_junction()` is new. It calls `self.lstat()` to get the stat result, and so it doesn't hit the underlying filesystem (unless `lstat()` does!). It would benefit from a review!

File: Lib/pathlib.py
Comment (barneygale) [inline Lib/pathlib.py]: This implementation of `_PathBase._scandir()` is new, and replaces a comment that alluded to its future addition. `_scandir()` is used by `glob()` and `walk()`. 

The `list()` call is necessary because `iterdir()` doesn't raise until you start iterating. The `iterdir()` behaviour was reported as a bug that we closed, but it could be revisited. See #78722.

File: Lib/pathlib.py
Comment (barneygale) [inline Lib/pathlib.py]: This implementation of `_PathBase.resolve()` is brand new, and probably the most complicated and interesting part of the patch.

It implements '..' segment and symlink resolution using POSIX rules, like `realpath --physical`. This makes it slightly incompatible with Windows rules.

It suppresses `UnsupportedOperation` exceptions from `absolute()`, which means returned paths may not be absolute. This is important for formats that support symlinks but not working directories (e.g. `.tar` archives).

It would benefit from careful review!

**Comment by ap--** [comment]
That sounds great. It would be a lot simpler to depend on an official backport 😊

**Comment by barneygale** [comment]
@merwok would you have the time/inclination to review this, do you think? No problem if not - I can put a message in the core dev forum asking for reviewers. Thank you!

File: Lib/pathlib.py
Comment (barneygale) [reply Lib/pathlib.py]: @eryksun would you have time to review this? It's an implementation of `_PathBase.resolve()` (~`os.path.realpath()`) that calls `p.absolute()` and `p.readlink()` under the hood, without performing any filesystem access of its own. This could be used by subclasses of `_PathBase` that need symlink support, like a `TarPath` class for tarballs.

File: Lib/pathlib.py
Comment (barneygale) [reply Lib/pathlib.py]: Alternatively we could make `_PathBase.is_junction()` immediately raise `UnsupportedOperation`, but it would make it the only `is_*()` method that doesn't call through to `stat()`.

**Comment by merwok** [comment]
I think someone else would be better suited to review the correctness of the new functions.

The name is good and the graph you had on discuss to explain pathlike vs pathbase was also very helpful, maybe a version of it (as image or text) could make it to the docs!

File: Lib/pathlib.py
Comment (AA-Turner) [inline Lib/pathlib.py]: I'm not sure this fully captures the function of this class -- after all, ``Path`` inherits from it, and therefore ``_PathBase`` may be used for concrete file systems, as well as "virtual".

File: Lib/pathlib.py
Comment (AA-Turner) [inline Lib/pathlib.py]:
```suggestion
    __fspath__ = None  # virtual paths have no local file system representation
```

File: Lib/pathlib.py
Comment (AA-Turner) [inline Lib/pathlib.py]: Minor: Is there a reason we can't import ``IO_REPARSE_TAG_MOUNT_POINT`` into the module scope along with the other stat names?

Also: is this a known idiom on Unix systems? The documentation for [``st_reparse_tag``](https://docs.python.org/3.13/library/os.html#os.stat_result.st_reparse_tag) and [``IO_REPARSE_TAG_*``](https://docs.python.org/3.13/library/stat.html#stat.IO_REPARSE_TAG_SYMLINK) both fall under sections marked as Windows-only. If these are common Unix idioms, we should update the documenatation for ``os`` and ``stat``.

File: Lib/pathlib.py
Comment (AA-Turner) [reply Lib/pathlib.py]: This could do with a comment explaining the need for ``nullcontext`` (as the contract is to return a scandir-like context manager)

Comment (AA-Turner) [reply Lib/pathlib.py]: > The `list()` call is necessary because `iterdir()` doesn't raise until you start iterating.

I think this ought be documented or reconsidered, as e.g. [``os.scandir``](https://docs.python.org/3.13/library/os.html#os.scandir) is explicitly noted as having a performance boost over ``listdir`` (& the usual memory concerns, etc). Would it be possible to call ``next`` once, and then store that item for yielding later (perhaps with ``itertools.chain.from_iterable``)?

File: Lib/pathlib.py
Comment (AA-Turner) [inline Lib/pathlib.py]: For a more accurate docstring (virtual file systems may have no notion of a CWD)
```suggestion
        """Return an absolute version of this path
        No normalization or symlink resolution is performed.
```

File: Lib/pathlib.py
Comment (AA-Turner) [inline Lib/pathlib.py]: Is there always a canonical path for virtual file systems? (E.g. on a geographically replicated storage system, is US-1 or GB-1 the 'canonical' home of an object?)

File: Lib/pathlib.py
Comment (AA-Turner) [inline Lib/pathlib.py:1237]: See comment above on notion of a CWD -- ought this be "UnsupportedOperation"?

File: Lib/pathlib.py
Comment (AA-Turner) [inline Lib/pathlib.py]: Perhaps a bit pedantic, but again on the point that we shouldn't assume contracts that would apply to a normal (Unix-biased) file system
```suggestion
        """Return a new path pointing to expanduser('~').
```

File: Lib/pathlib.py
Comment (AA-Turner) [inline Lib/pathlib.py:1243]: This will change error messages from ``UnsupportedOperation("Path.owner() is unsupported on this system")`` to ``UnsupportedOperation("Path.owner()")`` -- is the information that the failure is system dependent useful here // will it be missed?

File: Lib/test/test_pathlib.py
Comment (AA-Turner) [inline Lib/test/test_pathlib.py:1]: Perhaps we should also test that the docstrings of ``_PathBase`` and ``Path`` remain synchronised (for methods in the set that we want this to hold true)?

**Review by AA-Turner — COMMENTED** [review]
Prompted by your plea for reviewers: I haven't looked at ``_PathBase.resolve``, though, sorry.

A

File: Lib/pathlib.py
Comment (barneygale) [reply Lib/pathlib.py]: > Minor: Is there a reason we can't import `IO_REPARSE_TAG_MOUNT_POINT` into the module scope along with the other stat names?

It's only available on Windows

> Also: is this a known idiom on Unix systems? The documentation for [`st_reparse_tag`](https://docs.python.org/3.13/library/os.html#os.stat_result.st_reparse_tag) and [`IO_REPARSE_TAG_*`](https://docs.python.org/3.13/library/stat.html#stat.IO_REPARSE_TAG_SYMLINK) both fall under sections marked as Windows-only. If these are common Unix idioms, we should update the documenatation for `os` and `stat`.

No; it's a purely Windows idiom. On other platforms, the `.st_reparse_tag` lookup raises an AttributeError, which is caught below and the method returns `False`.

File: Lib/pathlib.py
Comment (barneygale) [reply Lib/pathlib.py]: :+1: I'll do that if we decide against fixing #78722. My preference is still to adjust `iterdir()` rather than work around the problem in `_scandir()`

File: Lib/pathlib.py
Comment (barneygale) [reply Lib/pathlib.py]: Junctions seem to be a Windows-only feature, whereas other types of file (directories, symlinks, mountpoints) have analogues in some virtual filesystems. And so it's not clear to me what `_PathBase.is_junction()` should do, really. Immediately `return False`? Raise `UnsupportedOperation`? The current code effectively always returns false, unless the user returns from `stat()` an object with a `st_reparse_tag` attribute set correctly, and is running on Windows.

File: Lib/pathlib.py
Comment (barneygale) [reply Lib/pathlib.py:1237]: If a user implements `absolute()`, then `Path('.').absolute()` is guaranteed to return the working directory. If they don't implement `absolute()`, the exception from the default implementation of `absolute()` will be raised and propagated.

File: Lib/pathlib.py
Comment (barneygale) [reply Lib/pathlib.py:1243]: It might be missed, yes. I could define a method like this and use it to raise exceptions - does that seem sensible to you?

```python
    def _unsupported(self, method):
        msg = f"{type(self).__name__}.{method} is unsupported"
        if isinstance(self, Path):
            msg += " on this system"
        raise UnsupportedOperation(msg)
```

File: Lib/test/test_pathlib.py
Comment (barneygale) [reply Lib/test/test_pathlib.py:1]: Done!

File: Lib/pathlib.py
Comment (barneygale) [reply Lib/pathlib.py]: I've revised this code - `is_junction()` now ignores the result of `lstat()` and returns false. I _think_ this is a good way forward, but I'm not sure.

File: Lib/pathlib.py
Comment (gst) [inline Lib/pathlib.py]: why this hard-coded limit ? 

I mean why not keep a list/set of already resolved link path here and check if we see one already existing ?

Comment (barneygale) [reply Lib/pathlib.py]: > why this hard-coded limit ?

Simply because that's what Linux does! 

https://github.com/torvalds/linux/blob/4fb0dacb78c6a041bbd38ddd998df806af5c2c69/include/linux/namei.h#L13

https://github.com/torvalds/linux/blob/4fb0dacb78c6a041bbd38ddd998df806af5c2c69/fs/namei.c#L1728-L1729

> I mean why not keep a list/set of already resolved link path here and check if we see one already existing ?

It looks like it's not particularly easy to detect symlink loops: https://unix.stackexchange.com/questions/99159/is-there-an-algorithm-to-decide-if-a-symlink-loops

File: Lib/pathlib.py
Comment (AA-Turner) [reply Lib/pathlib.py:1243]: Yep, seems a good compromise.

File: Lib/pathlib.py
Comment (AA-Turner) [reply Lib/pathlib.py:1237]: > is guaranteed to return the working directory

Fair enough, I suppose ``absolute()`` needs to handle what the CWD would be for file systems where that doesn't fully map across -- as you say perhaps by erroring.

File: Lib/pathlib.py
Comment (AA-Turner) [reply Lib/pathlib.py]: I think this might be overcomplicating it -- for the default implementation, perhaps just return ``False``, as junctions are clearly fairly domain-specific. The current method has a dollop of code just to say "no" or to raise an error!

File: Lib/pathlib.py
Comment (AA-Turner) [reply Lib/pathlib.py]: I agree fixing iterdir at the source would be cleaner

File: Lib/pathlib.py
Comment (ofek) [reply Lib/pathlib.py]: I'm not sure if the solution generalizes but this is how Hatchling prevents loops https://github.com/pypa/hatch/blob/hatchling-v1.18.0/backend/src/hatchling/builders/utils.py#L21-L31

`setuptools` has similar logic

File: Lib/pathlib.py
Comment (barneygale) [reply Lib/pathlib.py]: I think I agree, but just to be clear: that would make `is_junction()` the only `_PathBase` method (excepting `PurePath` methods) that does not raise `UnsupportedOperation` by default. A worthwhile compromise, given it's such an edge case?

edit: wrong! `resolve()` also doesn't raise by default. That makes me feel better about immediately returning false.

File: Lib/pathlib.py
Comment (gst) [reply Lib/pathlib.py]: > Simply because that's what Linux does!

Python is not Linux only ! ;)

```python
# before the while loop:
links_seen = set()

# then here:
s = str(link)
if s in links_seen:
    raise OSError(...)
links_seen.add(s)
```

?

Comment (barneygale) [reply Lib/pathlib.py]: It's a bit more subtle than simply bailing out if we see the same stat twice. This is a symlink loop by some definitions, but `resolve()` should be able to handle it as `realpath` does:

```sh
barney@acorn ~/projects/cpython/testdir/c $ ln -s . a
barney@acorn ~/projects/cpython/testdir/c $ ln -s . b
barney@acorn ~/projects/cpython/testdir/c $ realpath a/b/a/b/a/b
/home/barney/projects/cpython/testdir/c
```

Similarly, `glob.glob()` and `pathlib.Path.glob()` will expand these paths up to 40 levels:

```python
>>> g = pathlib.Path().glob('**/', follow_symlinks=True)
>>> next(g)
PosixPath('.')
>>> next(g)
PosixPath('b')
>>> next(g)
PosixPath('b/b')
>>> next(g)
PosixPath('b/b/b')
>>> next(g)
PosixPath('b/b/b/b')
```

(I suppose this generator will have around 2^40 items)

Also, this is rather performance-sensitive code, and I want to optimize for the common case where we _don't_ need to raise `ELOOP`. Counting symlinks, as Linux does, is about the cheapest implementation possible. It's performance sensitive because any user subclass of `PathBase` that supports symlinks will need to call `resolve()` in its most fundamental methods, like `stat()`, `iterdir()` and `open()`. For example, the implementation of `DummyPath.iterdir()` in the pathlib tests reads:

```python
    def iterdir(self):
        path = str(self.resolve())
        if path in self._files:
            raise NotADirectoryError(errno.ENOTDIR, "Not a directory", path)
        elif path in self._directories:
            return (self / name for name in self._directories[path])
        else:
            raise FileNotFoundError(errno.ENOENT, "File not found", path)
```

`TarPath` has the same requirement and needs a sprinkling of `self.resolve()`, which can slow it down considerably.

Comment (barneygale) [reply Lib/pathlib.py]: I'm struggling to dig out macOS docs, but [this page](https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man2/open.2.html) says:

```
     [ELOOP]            Too many symbolic links are encountered in translating
                        the pathname.  This is taken to be indicative of a
                        looping symbolic link.
```

So I think Mac uses a similar algorithm.

File: Lib/pathlib.py
Comment (barneygale) [reply Lib/pathlib.py]: Thanks for your help with this @AA-Turner. I've made `is_junction()` immediately return false.

File: Lib/pathlib.py
Comment (barneygale) [reply Lib/pathlib.py]: Resolving this thread as I've made `_PathBase.is_junction()` immediately return false, which avoids all the Windows-specific stuff.

File: Lib/pathlib.py
Comment (barneygale) [reply Lib/pathlib.py]: I've adjusted the exception message to read "Too many symbolic links in path", and stored the limit as `pathlib._MAX_SYMLINKS`.

File: Lib/pathlib.py
Comment (AA-Turner) [inline Lib/pathlib.py]: Minor: should these just be ``self._unsupported("...")``? (Removing the ``return`` statement)

There's no functional difference, but seeing a ``return`` hints to a reader that there is a returned value, whereas these functions unconditionally raise an exception.

A

**Comment by barneygale** [comment]
For historical reasons, `Path.resolve()` raises `RuntimeError` rather than `OSError` when a symlink loop is encountered, and does so even when `strict=False`. I've just logged an issue about it: #109187

The implementation of `_PathBase.resolve()` in this PR does the right thing: it raises `OSError(ELOOP)` when `strict=True`, and suppresses the exception when `strict=False`.

But that makes the two `resolves()`s slightly incompatible. I _think_ we should fix #109187 in 3.13, rather than baking the nonsensical `RuntimeError` stuff into `_PathBase.resolve()`.

File: Lib/pathlib.py
Comment (barneygale) [reply Lib/pathlib.py]: Thanks, I've adjusted this to remove the `return`s!

**Comment by RonnyPfannschmidt** [comment]
just stumbled uppon this and im wondering if a concept of a public Accessor of a path might make sense
where an accessor provides the implementation, then for example tarfile/zipfile paths could just have the accessor

additionally it would be trivial to have "roots" at a path or a "root" with a directory file descriptor/network protocol

**Comment by RonnyPfannschmidt** [comment]
those "accessors" where nothing more than aliases, after all the implementation as done had no power over what the pure path portion can do with the real world

a actualy "accessor" would be able to be tied to something like a directory file descriptor, or a path object, or a entirely different thing thats not a filesystem

and yes there is some need for synchronization between concrete path capabilities and the accessors,

however there is certain equivalence classes to keep in mind (Posix, Windows), so in a lot of cases my impression is that there will be more "accessors" than paths

**Comment by barneygale** [comment]
Intriguing. Could you sketch out the accessor API?

**Comment by merwok** [comment]
(in the forums rather than a PR discussion please)

**Comment by RonnyPfannschmidt** [comment]
i can dedicate some time to it later

a basic concept  of the accessor version ought to consider something like:

* windows paths
* posix paths
* unc paths
* directory file descriptor paths (both secured and unsecured)
* paths that pretend to be a chroot on top of another path/pathlike
* archives
* ssh/scp
* urls/http
* distinctions between readonly and readwrite accessors
* capabilities of accessors (like users/perms/xattrs)
* wrappers (like union fs) to support interesting testing 
   (imagine a path that maps onto the real fs but pretends certain files are deleted/replaced) 
   
   additionally for purely virtual accessors it might be interesting if parts of the path could be objects (like dates), but thats currently out of scope

**Comment by barneygale** [comment]
To summarise my thoughts on the above: in many cases this PR helps bring those things about, and I don't think it hinders any of those things. But let's chat more in a forum thread.

What's left in this PR:

- Land #109192 to align handling of symlink loops in `Path.resolve()`'s with this implementation of `_PathBase.resolve()`, and with `os.path.realpath()`
- Get review/approval for this `_PathBase.resolve()` implementation.

On the latter point, there are still two subtle and related problems:

- `_PathBase.resolve(strict=True)` can raise `OSError` with the wrong values (errno, path). For example, `resolve('/etc/hosts/foo')` should raise `IsADirectoryError(path='/etc/hosts')`, but instead raises `FileNotFoundError(path='/etc/hosts/foo')`
- To avoid an infinite loop, the implementation of `_PathBase.readlink()` must not call `self.resolve()`, and so users must supply a path with no symlinks, except in the final position. This is a bit different to `Path.readlink()`.

Both of these issues could be solved if we can provide some way to `stat()` a path without resolving _any_ symlinks. The `resolve()` method would call this `stat()` method first, and only call `readlink()` if the stat result indicates a symlink. A couple of options spring to mind:

- We could allow `_PathBase.stat(follow_symlinks=None)`, meaning "where possible, don't follow _any_ symlinks in the path; the path I've provided is the direct physical path". But it's not particularly obvious, and `follow_symlinks=None` means something different in `glob()` and `rglob()`.
- We could add some other keyword-only argument to `_PathBase.stat()` with the same effect.
- We could add a new method -- `direct_stat()`, `stat1()`, `dstat()`?

**Comment by barneygale** [comment]
Bah, pushed a commit labelled "WIP". In fact it contains a solution to the above problem: stash a `_resolving` flag on path objects, which means `resolve()` can safely call `stat()` and `readlink()` without hitting infinite recursion. No weird APIs for users.

**Comment by barneygale** [comment]
Hey @AA-Turner and @gst, thank you for your earlier reviews, very much appreciated. Please could you give it another pass, if you can spare the time?

I feel OK about `_PathBase.resolve()`: it's entirely private for the timebeing, and I intend to add further tests and optimization before we drop the underscore from `_PathBase`, so if we can't spot any clear defects I suspect it's fine to land? Grateful for your guidance!

**Review by gst — APPROVED** [review]

File: Lib/pathlib.py
Comment (AA-Turner) [inline Lib/pathlib.py]: Could this be extracted to a module level ``_split_path`` to avoid redefining this function on each call to ``resolve()``?

File: Lib/pathlib.py
Comment (AA-Turner) [inline Lib/pathlib.py]: Perhaps add a note on why we use the string representation here for the cache keys rather than using e.g. ``_drv, _root, _tail``?

File: Lib/pathlib.py
Comment (AA-Turner) [inline Lib/pathlib.py]: I don't think ``path_str`` can ever be ``None``, so potentially this could use try-except? I suppose it depends on if we expect to hit this cache often or not
```suggestion
                    try:
                        st = stat_cache[path_str]
                    except KeyError:
                        st = stat_cache[path_str] = path.stat(follow_symlinks=False)
```

File: Lib/pathlib.py
Comment (AA-Turner) [inline Lib/pathlib.py]: The same point as ``stat_cache``:
```suggestion
                        try:
                            target = target_cache[path_str]
                        except KeyError:
                            target = target_cache[path_str] = path.readlink()
```

File: Lib/pathlib.py
Comment (barneygale) [reply Lib/pathlib.py]: I'll switch to using the path object itself as a cache key. In future we might set `__hash__ = None` in `PathBase` and force users to write their own `__hash__()` if they want their path objects to be hashable, at which point using the `str(path)` as a cache key will be easier to justify.

File: Lib/pathlib.py
Comment (barneygale) [reply Lib/pathlib.py]: The cache only helps with symlink loops or upward-pointing symlinks, which are an unusual case. Cache _misses_ are much more common in practice.

File: Lib/pathlib.py
Comment (barneygale) [reply Lib/pathlib.py]: I've added a `PathBase._split_stack()` method rather than a top-level function. Not too sure about naming/placement, let me know if you think module-level makes more sense.

File: Lib/pathlib.py
Comment (AA-Turner) [inline Lib/pathlib.py]: I found this a little hard to follow due to the name reuse. Perhaps some explanatory comments about what's going on?

File: Lib/pathlib.py
Comment (AA-Turner) [inline Lib/pathlib.py]: As far as I can tell ``readlink._unsupported = True`` is never set to ``False``? Perhaps this could also switch to ``readlink._supported = False``, the double- and triple-negatives are a bit confusing at times.

**Review by AA-Turner — APPROVED** [review]
Thanks! Just two notes. I think ``_split_stack`` makes sense a method.

A

File: Lib/pathlib.py
Comment (barneygale) [reply Lib/pathlib.py]: It would be unset if a user subclasses `PathBase` and adds their own `readlink()` method. I've added a comment and renamed it to `_supported` - does it make more sense now?

**Comment by barneygale** [comment]
Thank you so much for the reviews, everyone :-)

**Comment by AA-Turner** [comment]
Congratulations Barney, a lot of hard work on your part paid off!

A

File: Lib/pathlib.py
Comment (AA-Turner) [reply Lib/pathlib.py]: I was confused as to why ``Path`` didn't change this -- but of course ``os.path.realpath`` is used instead!

A

**Pull request merged**
